swap/swap.go

```diff
@@ -256,13 +256,46 @@ func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
 }
 
 //GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+func (s *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	var peerBalance int64
+	var err error
+
+	// load balance from memory
+	peerBalance, keyExists := s.balances[peer]
+	if !keyExists {
+		// if not in memory, try to load balance from disk
+		err = s.stateStore.Get(peer.String(), &peerBalance)
+	}
+	return peerBalance, err
+}
+
+//GetAllBalances returns the balances for all known peers
+func (s *Swap) GetAllBalances() (map[enode.ID]int64, error) {
+	balances := make(map[enode.ID]int64)
+
+	// load balances from disk
+	peerIDs, err := s.stateStore.Keys()
+	if err != nil {
+		return nil, err
 	}
-	return 0, errors.New("Peer not found")
+
+	for _, peerIDString := range peerIDs {
+		peerID := enode.HexID(peerIDString)
+		peerBalance, err := s.GetPeerBalance(peerID)
+		if err != nil {
+			return nil, err
+		}
+		balances[peerID] = peerBalance
+	}
+
+	// add balances from memory
+	for peerID, peerBalance := range s.balances {
```

Comment:
One variation could be to initialize `balances` right at the top to be `s.balances`:

`balances := s.balances`

Then, when we read from file, we would have to check if the keys is already there and not overwrite.
What do you think is best?

Also, we may have to `lock` while reading the balances from memory

Reply:
This comment made me see that this function had redundant logic, since the inner `GetPeerBalance` call already checked both memory and disk, and applied the locks as well.

I've now refactored this function into one with 2 steps: get a list of SWAP peers, and then get the balance for each of them. Both of these steps deal with the locks and memory/disk access, and could be called independently.

While I understand it might not be optimal in terms of performance (while getting the list of peers we could be getting their balances as well) I do think it's the cleanest approach, so as to segregate responsibilities.